transport-classes-io_uring/src/main/java/io/netty/channel/uring/IoUringBufferRing.java

```diff
@@ -0,0 +1,452 @@
+        PlatformDependent.putShortOrdered(tailFieldAddress, newTail);
+        hasSpareBuffer = true;
+    }
+
+    int entries() {
+        return entries;
+    }
+
+    short bufferGroupId() {
+        return bufferGroupId;
+    }
+
+    int chunkSize() {
+        return chunkSize;
+    }
+
+    boolean isFull() {
+        return nextIndex == entries;
+    }
+
+    long address() {
+        return ioUringBufRingAddr;
+    }
+
+    void close() {
+        Native.ioUringUnRegisterBufRing(ringFd, ioUringBufRingAddr, 4, 1);
+        for (ByteBuf byteBuf : userspaceBufferHolder) {
+            if (byteBuf != null) {
+                byteBuf.release();
+            }
+        }
+    }
+
+    class UserspaceIoUringBuffer extends AbstractReferenceCountedByteBuf {
+
+        private final short bid;
+
+        private final ByteBuf userspaceBuffer;
+
+        protected UserspaceIoUringBuffer(int maxCapacity, short bid, ByteBuf userspaceBuffer) {
+            super(maxCapacity);
+            this.bid = bid;
+            this.userspaceBuffer = userspaceBuffer;
+        }
+
+        @Override
+        protected void deallocate() {
+            userspaceBufferHolder[bid].release();
+            recycleBuffer(bid);
+        }
+
+        @Override
+        protected byte _getByte(int index) {
+            return userspaceBuffer.getByte(index);
+        }
+
+        @Override
+        protected short _getShort(int index) {
+            return userspaceBuffer.getShort(index);
+        }
+
+        @Override
+        protected short _getShortLE(int index) {
+            return userspaceBuffer.getShortLE(index);
+        }
+
+        @Override
+        protected int _getUnsignedMedium(int index) {
+            return userspaceBuffer.getUnsignedMedium(index);
+        }
+
+        @Override
+        protected int _getUnsignedMediumLE(int index) {
+            return userspaceBuffer.getUnsignedMediumLE(index);
+        }
+
+        @Override
+        protected int _getInt(int index) {
+            return userspaceBuffer.getInt(index);
+        }
+
+        @Override
+        protected int _getIntLE(int index) {
+            return userspaceBuffer.getIntLE(index);
+        }
+
+        @Override
+        protected long _getLong(int index) {
+            return userspaceBuffer.getLong(index);
+        }
+
+        @Override
+        protected long _getLongLE(int index) {
+            return userspaceBuffer.getLongLE(index);
+        }
+
+        @Override
+        protected void _setByte(int index, int value) {
+            userspaceBuffer.setByte(index, value);
+        }
+
+        @Override
+        protected void _setShort(int index, int value) {
+            userspaceBuffer.setShort(index, value);
+        }
+
+        @Override
+        protected void _setShortLE(int index, int value) {
+            userspaceBuffer.setShortLE(index, value);
+        }
+
+        @Override
+        protected void _setMedium(int index, int value) {
```

Comment:
See https://github.com/netty/netty/pull/14287/ after addressing wrapping it differently

Reply:
Thanks bud :rocket: